Feature/devotion detail

jdwitten opened the pull request:
- This gets us started with a devotion detail design and a custom devotion detail table view cell.
- Also, records if a devotion has been read and displays a dot if it is unread on the table view

Commit:
Add a settings state and the first stage of scheduling notifications

Commit:
Merge branch 'master' into feature/devotion-notifications

# Conflicts:
#	iOS Client/Actions/MBActions.swift
#	iOS Client/Store/MBStore.swift

Commit:
Create a devotion controller and schedule notifications and handle received notifications

Commit:
Style devotion detail and mark dates of devotions that have been read in a file

AlexRamey commented on iOS Client/DevotionsController/DevotionDetailViewController.swift:
```diff
@@ -0,0 +1,65 @@
+            let monthInt = Formatters.calendar?.component(.month, from: date),
+            let day = Formatters.calendar?.component(.day, from: date) {
+            monthLabel.text = Formatters.getMonth(fromInt: monthInt)
+            dayLabel.text = String(describing: day)
```
that's cool, never seen String describing before

AlexRamey replied:
maybe worth saving the formatted month string and the formatted day string as lazy properties on the devotion class so we don't have to burden this function with these details

AlexRamey commented on iOS Client/DevotionsController/DevotionDetailViewController.swift:
```diff
@@ -0,0 +1,65 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        configureBackButton()
```
maybe we should subclass UIViewController and override viewDidLoad to configure the back button for us since we do it more than once? We could also do the store subscribe and unsubscribe in viewWillAppear and viewWillDisappear since we do that in all our VCs. Thinking out loud here, I can handle it in a later refactor if u think it's a good idea

AlexRamey commented on iOS Client/DevotionsController/DevotionTableViewCell.swift:
```diff
@@ -0,0 +1,28 @@
+        monthLabel.text = month
+        verseLabel.text = verse
+        
+        readNotifier.layer.cornerRadius = readNotifier.bounds.width / 2
```
very much looking forward to seeing this in action

AlexRamey commented on iOS Client/DevotionsController/MBDevotionsViewController.swift:
```diff
@@ -12,11 +12,14 @@ import ReSwift
         tableView.dataSource = self
+        tableView.rowHeight = UITableViewAutomaticDimension
+        
+        tableView.register(UINib(nibName: "DevotionTableViewCell", bundle: nil), forCellReuseIdentifier: "DevotionTableViewCell")
```
it would be good to make a constant called `reuseIdentifier` as a private property in this class

AlexRamey replied:
and use it here and in tableView cellForRowAt

AlexRamey commented on iOS Client/DevotionsController/MBDevotionsViewController.swift:
```diff
@@ -52,8 +55,15 @@ class MBDevotionsViewController: UIViewController, StoreSubscriber, UITableViewD
+        //swiftlint:enable force_cast
+        if let date = Formatters.devotionDateFormatter.date(from: devotion.devotion.date),
+            let day = Formatters.calendar?.components(.day, from: date).day,
+            let monthInt = Formatters.calendar?.component(.month, from: date) {
```
if we move this into the devotion class as a lazy property, we don't have to repeat these computations everywhere. Maybe they will have to be optional since there's a chance of failure, or maybe we should throw in `!` and fail fast since we theoretically control the data source

AlexRamey commented on iOS Client/TabController/AppCoordinator.swift:
```diff
@@ -62,4 +108,19 @@ class AppCoordinator: NSObject, Coordinator, UITabBarControllerDelegate, StoreSu
     }
+    
+    // MARK: - UNNotificationDelegate
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
```
this should be devotion coordinator's concern

AlexRamey commented on iOS Client/TabController/AppCoordinator.swift:
```diff
@@ -43,6 +44,51 @@ class AppCoordinator: NSObject, Coordinator, UITabBarControllerDelegate, StoreSu
+            dateComponents.minute = 30
+            dateComponents.year = calendar.component(NSCalendar.Unit.year, from: devotionDay)
+            dateComponents.month = calendar.component(NSCalendar.Unit.month, from: devotionDay)
+            dateComponents.day = calendar.component(NSCalendar.Unit.day, from: devotionDay)
```
this needs to be moved inside the devotion class

AlexRamey commented on iOS Client/Utilities/Formatters.swift:
```diff
@@ -0,0 +1,53 @@
+            return "October"
+        case 11:
+            return "November"
+        case 12:
```
default could be December

AlexRamey replied:
i have a feeling this functionality exists for us somewhere . . .

AlexRamey replied:
Try `DateFormatter().monthSymbols[monthInt-1]`

AlexRamey replied:
monthSymbols returns `["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]` for me in a playground

jdwitten replied:
Boom didn't know this existed nice

AlexRamey commented on iOS Client/TabController/AppCoordinator.swift:
```diff
@@ -43,6 +44,51 @@ class AppCoordinator: NSObject, Coordinator, UITabBarControllerDelegate, StoreSu
+            let trigger = UNCalendarNotificationTrigger(dateMatching: dateComponents, repeats: true)
+            let content = UNMutableNotificationContent()
+            content.title = devotion.verse
+            content.body = "Read your daily devotion from Mockingbird"
```
we should subclass UNMutableNotificationContent and do this setup in an `init`

AlexRamey commented on iOS Client/TabController/AppCoordinator.swift:
```diff
@@ -43,6 +44,51 @@ class AppCoordinator: NSObject, Coordinator, UITabBarControllerDelegate, StoreSu
+        promptForNotifications()
+    }
+    
+    func promptForNotifications() {
```
should be devotion coordinator concern, see other PR comments for suggested flow

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -22,6 +22,21 @@ class MBStore: NSObject {
+        return MBAppState(navigationState: MBNavigationState(), articleState: MBArticleState(), devotionState: MBDevotionState(), settingsState: MBSettingsState(notificationPermission: notificationPermission))
+    }
+    
+    static func saveState(state: AppState) {
```
this and loadState appear to be unusued

AlexRamey commented on iOS Client/Models/MBDevotion.swift:
```diff
@@ -9,9 +9,14 @@
     var verseText: String
 }
+
+struct LoadedDevotion: Codable, Detailable {
```
Hmmm, I don't think we should have a distinction between these two types

AlexRamey replied:
As per discussion, I was wrong about this point. We absolutely need two different structs. One that defines the contract between us and the external data source, and another that is our internal domain model.

They should be independent  (so our domain model should not wrap the contract model but just copy over its fields) so that we can insulate our app domain from changes to the external contract. For many changes, we can just update our conversion function that converts the contract model to our domain model.

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -96,35 +129,60 @@ class MBStore: NSObject {
-                    self.save(data, pathUrl) { error in
-                        if error != nil {
-                            print("Could not save devotions data to disk")
+    func markDevotionAsRead(date: String, completion: @escaping (Error?) -> Void) {
```
I think this could be done in a simpler way. Why de-couple the devotions from the read dates? Whether or not a devotion is `read` is a property of that devotion. It should be part of our MBDevotion model. 

When we first download the devotions data, we shouldn't blindly save it. We should get it from the external source (whether that's the internet or a file in the app bundle), deserialize it into our Model Objects, which can have extra fields beyond what's in the JSON like `isRead` that will default to `false` and then serialize the array of our model devotions into a file at the save path. 

The only issue is if we want to make the apps reload the devotions from the external source for some reason. In that case, we would need a helper function to merge what's new with what we have, but that's better in my opinion than maintaining dates separately and dealing with two different devotion models

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -96,35 +129,60 @@ class MBStore: NSObject {
+        do {
+            let (manager, path, _) = try urlPackage(forPath: "/\(fromPath)")
+            if !manager.fileExists(atPath: path), let file = Bundle.main.url(forResource: fromPath, withExtension: "json") {
+                    let data = try Data(contentsOf: file)
```
I miss the notion you had before of moving the data from the external source to a new location inside our document directory. I think it's important to decouple the data source from our internal store

AlexRamey reviewed (changes requested):
-----

AlexRamey commented:
---

Commit:
Refactor devotion read date save and read functions

Commit:
make table view cell constant

Commit:
Refactor some of the date formatting and devotion models

jdwitten commented:
This bad boy is ready for another review. I addressed most of your comments and reworked some of the devotions syncing. I realize it is a little bit hard to follow in MBStore but I think it will look much nicer if we get something like PromiseKit so we can handle any errors in one place

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -78,11 +78,65 @@ class MBStore: NSObject {
-    func syncDevotions(completion: @escaping ([MBDevotion]?, Error?) -> Void) {
-        loadDevotions { (devotions, devotionError) in
-            completion(devotions, devotionError)
+    func syncDevotions(completion: @escaping ([LoadedDevotion]?, Error?) -> Void) {
```
I think this is still too complex. So conceptually, the `sync` part of this seems to be syncing the devotions data stored locally with the readDates file. (There is some notion of syncing with the external store, but really that's a lazy fetch of the devotions data in case we don't have it locally)

Despite a potential performance hit, I propose we do away with the extra readDatesFile and just store the array of LoadedDevotion objects. This means that marking a devotion as read is reading out the entire array of all devotions, finding the one of interest and marking it as read, then saving the array back to disk. I think that's better than maintaining two files.

I then think this method could be decomposed into two. `getDevotions` that blindly reads LoadedDevotions from documents directory and synchronously returns them, and `syncDevotions` that pulls them from the external store, converts them to LoadedDevotions, and saves them to documents directory (as well as returns them in a completion block) as you've done here. 

Then, everywhere in our app we need devotions, we can just call `getDevotions` without needing a completion block. How the sync initially gets kicked off is up for discussion. Maybe our state will let us know if they've been loaded yet. Maybe getDevotions can just return an empty list and kick off a sync job that triggers a store action on completion. Not entirely sure, but I do think it's important to

1) eliminate the extra readDatesFile and
2) create a synchronous `getDevotions`

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -96,42 +150,69 @@ class MBStore: NSObject {
-                        if error != nil {
-                            print("Could not save devotions data to disk")
-                        }
+    func markDevotionAsRead(date: String, completion: @escaping (Error?) -> Void) {
```
why does this use a completion block? Can't it just return an error type directly since there's nothing async about it?

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -96,42 +150,69 @@ class MBStore: NSObject {
+        return (manager, path, pathUrl)
+    }
+    
     private func parse<T: Codable>(_ data: Data, _ resource: T.Type, _ completion: @escaping (T?, Error?) -> Void) {
```
why completion block? sync code should just return a value directly

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -96,42 +150,69 @@ class MBStore: NSObject {
     }
     
+    
+    private func read<T: Codable>(fromPath: String, _ resource: T.Type, completion: @escaping (T?, Error?) -> Void) {
```
why completion block? sync code  should just return a value directly

AlexRamey commented on iOS Client/DevotionsController/MBDevotionsViewController.swift:
```diff
@@ -64,5 +75,20 @@ class MBDevotionsViewController: UIViewController, StoreSubscriber, UITableViewD
+        let devotion = devotions[indexPath.row]
+        MBStore.sharedStore.dispatch(SelectedDevotion(devotion: devotion.devotion))
+        let store = MBStore()
+        store.markDevotionAsRead(date: devotion.devotion.date) { error in
```
This shouldn't have the look and feel of an async call.

AlexRamey commented on iOS Client/DevotionsController/MBDevotionsViewController.swift:
```diff
@@ -64,5 +75,20 @@ class MBDevotionsViewController: UIViewController, StoreSubscriber, UITableViewD
+            if error == nil {
+                store.syncDevotions { devotions, error in
+                    if error == nil, let loadedDevotions = devotions {
+                        MBStore.sharedStore.dispatch(LoadedDevotions(devotions: .loaded(data: loadedDevotions)))
```
sketchy to do this in a completion block b/c you don't know what thread you're on. This needs to be in a dispatch to the main thread (or, my vote is to make markDevotionAsRead just return an error type and then you can check the error and do this imperatively on the next line. No need for this completion block approach IMO since this is not async code)

AlexRamey commented on iOS Client/DevotionsController/MBDevotionsViewController.swift:
```diff
@@ -12,11 +12,15 @@ import ReSwift
         tableView.dataSource = self
+        tableView.rowHeight = UITableViewAutomaticDimension
+        
+        tableView.register(UINib(nibName: cellReusableId, bundle: nil), forCellReuseIdentifier: cellReusableId)
```
it's kind of a coincidence that the nibName is the same as the cellReusableId. I'd still hard-code the nibName so the reuse ID value can change without hecking up this code

AlexRamey replied:
`nibName` should be hard-coded in my opinion. The nibName and the cellReuseIdentifier should be independent of one another

Commit:
Simplify devotion synchronization code

Commit:
Add a getDevotions method that reads from file

jdwitten commented:
Alright let's give this another shot. I think I like how this one ended up. Here are some changes I made:

1. Got rid of sycWithReadDates and just save LoadedDevotions in document file
2. Added a getDevotions method that just reads blindly from the file
3. Made a bunch of methods synchronous without completion blocks (not sure why I thought they had to be async)
4. Remove markDevotionAsRead() from the store and just added a new method to just save some LoadedDevotions to file. So the reducer now marks devotion as read when it is selected and then all devotions are saved

AlexRamey commented on iOS Client/Utilities/Formatters.swift:
```diff
@@ -0,0 +1,26 @@
+    }()
+    
+    static func getMonth(fromInt: Int) -> String {
+        return Formatters.devotionDateFormatter.monthSymbols[fromInt-1]
```
👍

AlexRamey commented on iOS Client/DevotionsController/DevotionDetailViewController.swift:
```diff
@@ -0,0 +1,65 @@
+            let date = Formatters.devotionDateFormatter.date(from: devotion.date),
+            let monthInt = Formatters.calendar?.component(.month, from: date),
+            let day = Formatters.calendar?.component(.day, from: date) {
+            monthLabel.text = Formatters.getMonth(fromInt: monthInt)
```
Can be done in a later refactor, but I think It'd be cool to add monthName and monthDay (idk if those are the best names, maybe like displayDateMonthComponent displayDateDayComponent . . .) as lazily computed string properties on LoadedDevotion that encapsulate all this logic.

AlexRamey commented on iOS Client/DevotionsController/DevotionNotification.swift:
```diff
@@ -0,0 +1,26 @@
+        self.sound = UNNotificationSound.default()
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
```
Hmm, wonder what the deal with this is? It breaks if we delete it?

AlexRamey replied:
the swift docs are kind of unclear. They say all subclasses have to provide an implementation or required initializers but also that `You do not have to provide an explicit implementation of a required initializer if you can satisfy the requirement with an inherited initializer.` I don't know what that means . . .

jdwitten replied:
Yeah I'm a little confused about this. I don't think this ever actually would get called but it would be nice to not include it

AlexRamey commented on iOS Client/DevotionsController/DevotionTableViewCell.swift:
```diff
@@ -0,0 +1,28 @@
+    @IBOutlet weak var readNotifier: UIImageView!
+    
+    
+    func configure(day: String, month: String, verse: String, read: Bool) {
```
a minor note, but this could just accept one parameter, the LoadedDevotion

jdwitten replied:
nice I like that

AlexRamey commented on iOS Client/DevotionsController/DevotionsCoordinator.swift:
```diff
@@ -25,13 +26,53 @@ class DevotionsCoordinator: NSObject, Coordinator, StoreSubscriber {
+            }
+        } else {
+            MBStore.sharedStore.dispatch(LoadedDevotions(devotions: Loaded.loaded(data: devotions)))
+        }
```
good stuff.

AlexRamey commented on iOS Client/DevotionsController/DevotionsCoordinator.swift:
```diff
@@ -25,13 +26,53 @@ class DevotionsCoordinator: NSObject, Coordinator, StoreSubscriber {
+        if devotions.count == 0 {
+            store.syncDevotions { syncedDevotions, error in
+                if error != nil {
+                    MBStore.sharedStore.dispatch(LoadedDevotions(devotions: Loaded.error))
```
To be safe, these two dispatches should be in a dispatch.main.async block to guarantee that they occur on the main thread, since that is a requirement of ReSwift. In this case, I think it would be on the main thread b/c our client method to load the devotions is probably synchronous, but if we changed that out to be a network call then we might get some random undefined behavior stemming from these dispatches occurring on a background thread.

jdwitten replied:
good call I should put that first dispatch in an async block

AlexRamey commented on iOS Client/DevotionsController/DevotionsCoordinator.swift:
```diff
@@ -25,13 +26,53 @@ class DevotionsCoordinator: NSObject, Coordinator, StoreSubscriber {
+                }
+            }
+        } else {
+            MBStore.sharedStore.dispatch(LoadedDevotions(devotions: Loaded.loaded(data: devotions)))
```
this one is fine, def on the main thread

AlexRamey commented on iOS Client/DevotionsController/DevotionsCoordinator.swift:
```diff
@@ -43,4 +84,19 @@ class DevotionsCoordinator: NSObject, Coordinator, StoreSubscriber {
+    
+    // MARK: - UNNotificationDelegate
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+        MBStore().syncDevotions { devotions, error in
```
this can just be a synchronous getDevotions

AlexRamey commented on iOS Client/DevotionsController/DevotionsCoordinator.swift:
```diff
@@ -25,13 +26,53 @@ class DevotionsCoordinator: NSObject, Coordinator, StoreSubscriber {
+        }
+    }
+    
+    func promptForNotifications() {
```
where does this get called?

AlexRamey replied:
maybe it could be called from a newState method when the devotions get loaded, that way we know we can use `getDevotions` in the body of prompt and schedule. Also might be worth throwing a bool in UserDefaults called `areNotificationsScheduled` or something so we can do nothing if it's already taken care of

jdwitten replied:
whoops this used to get called in the DevotionsCoordinator start method but must have gotten lost in the shuffle. Yeah I like having a field in UserDefaults I can add that

jdwitten replied:
I'm going to keep the promptForNotifications call after we sync devotions on the first start up of the devotions coordinator. It's not going to do anything if we have already scheduled or if the user has already been prompted. Not sure exactly where it would go in newState because we wouldn't want to schedule every time we got a new state. Could add a new property on DevotionState that said whether or not we have scheduled yet. I'll leave it for now and we can refactor if necessary.

AlexRamey replied:
what you did is good

AlexRamey commented on iOS Client/DevotionsController/DevotionsCoordinator.swift:
```diff
@@ -25,13 +26,53 @@ class DevotionsCoordinator: NSObject, Coordinator, StoreSubscriber {
-                MBStore.sharedStore.dispatch(LoadedDevotions(devotions: Loaded.loaded(data: devotions)))
+        center.requestAuthorization(options: [.alert, .badge, .sound]) { granted, error in
+            if granted {
+                MBStore().syncDevotions { devotions, error in
```
this can be a `get`

jdwitten replied:
good call

AlexRamey commented on iOS Client/DevotionsController/MBDevotionsViewController.swift:
```diff
@@ -12,11 +12,17 @@ import ReSwift
-    var devotions: [MBDevotion] = []
+    var store: MBStore!
+    var devotions: [LoadedDevotion] = []
+    var cellReusableId: String = "DevotionTableViewCell"
```
i've usually seen `reuseIdentifier`, but I'm not gonna split hairs over it (and the value would be something more like `DevotionCellReuseIdentifier`)

AlexRamey commented on iOS Client/DevotionsController/MBDevotionsViewController.swift:
```diff
@@ -52,8 +58,15 @@ class MBDevotionsViewController: UIViewController, StoreSubscriber, UITableViewD
+            let day = Formatters.calendar?.components(.day, from: date).day,
+            let monthInt = Formatters.calendar?.component(.month, from: date) {
+            cell.configure(day: String(describing: day), month: Formatters.getMonth(fromInt: monthInt), verse: devotion.devotion.verse, read: devotion.read)
+
```
whitespace. Could be cool to just say cell.configure(devotion) and let it handle everything. If it encounters any problems with dates, it can use empty fields. As it stands now, if a date error were to occur, the cell would have old values still in it from its previous position in the table view

AlexRamey replied:
i've seen some quotes before in a book called Clean Code that a function should accept at most two parameters (kind of ridiculous, but something to strive for nonetheless 🐘 )

AlexRamey commented on iOS Client/DevotionsController/MBDevotionsViewController.swift:
```diff
@@ -64,5 +77,17 @@ class MBDevotionsViewController: UIViewController, StoreSubscriber, UITableViewD
+       let devotion = devotions[indexPath.row]
+        MBStore.sharedStore.dispatch(SelectedDevotion(devotion: devotion))
+        do {
+            try store.saveDevotions(devotions: devotions)
```
before this, where does the read property on the devotion actually get toggled to 'true' ?

jdwitten replied:
So that happens in the reducer during the SelectedDevotion action

AlexRamey replied:
Hmm, I am OK with this for now, but I think it needs to be clearer in an upcoming refactor when we mark a devotion as read

The concerning thing is that the call to dispatch has a side effect of updating this view controller's devotions property that the next lines rely on.

I think it would be clearer to only do the dispatch here, then in `newState` check if the selected devotion in the devotions property is currently marked as false and, if so, call a helper to do the save

jdwitten replied:
I agree that it isn't ideal that the view controllers devotions rely on that dispatched action. Although, I don't think saving in the reducer is a good idea. Reducers shouldn't have any side effects like saving data or fetching data, they should just take a state and an action and return a new state, nothing else. I'll try to think of a better solution but might be good to just get this in.

AlexRamey commented on iOS Client/DevotionsController/MBDevotionsViewController.swift:
```diff
@@ -64,5 +77,17 @@ class MBDevotionsViewController: UIViewController, StoreSubscriber, UITableViewD
+        do {
+            try store.saveDevotions(devotions: devotions)
+        } catch {
+            //There was an error saving devotion as read so reverse
```
nice handling of this case.

AlexRamey commented on iOS Client/HTTPClient/MBClient.swift:
```diff
@@ -108,9 +108,9 @@ class MBClient: NSObject {
+    func getJSONFile(name: String, completion: @escaping (Data?, Error?) -> Void ) {
         do {
-            if let file = Bundle.main.url(forResource: "devotions", withExtension: "json") {
+            if let file = Bundle.main.url(forResource: name, withExtension: "json") {
```
nice

AlexRamey commented on iOS Client/HTTPClient/MBClient.swift:
```diff
@@ -108,9 +108,9 @@ class MBClient: NSObject {
     }
     
-    func getDevotionsWithCompletion(completion: @escaping (Data?, Error?) -> Void ) {
+    func getJSONFile(name: String, completion: @escaping (Data?, Error?) -> Void ) {
```
so, this is also synchronous, but I'm OK with it exposing an async interface since we could conceivably change it to happen in the background. This hints to callers of this code that their completion block may not run on the main thread

jdwitten replied:
Yeah I'm going to keep it like this just in case in the future we have to get this file over the network

AlexRamey commented on iOS Client/DevotionsController/DevotionsCoordinator.swift:
```diff
@@ -25,13 +26,53 @@ class DevotionsCoordinator: NSObject, Coordinator, StoreSubscriber {
+                        print(error)
+                    } else if devotions != nil {
+                        center.delegate = self
+                        self.scheduleNotifications(devotions: devotions!.map{$0.devotion})
```
this code seems out-dated, since the LoadedDevotions no longer contain an embedded MBDevotion

AlexRamey commented on iOS Client/DevotionsController/DevotionsCoordinator.swift:
```diff
@@ -25,13 +26,53 @@ class DevotionsCoordinator: NSObject, Coordinator, StoreSubscriber {
+        }
+    }
+    
+    func scheduleNotifications(devotions: [MBDevotion]) {
```
should accept an array of LoadedDevotions

jdwitten replied:
yep forgot about that one

AlexRamey commented on iOS Client/Models/MBDevotion.swift:
```diff
@@ -9,9 +9,46 @@
     var verse: String
     var verseText: String
+    
+    var dateComponentsForNotification: DateComponents? {
```
this behavior will need to be moved down into the LoadedDevotion class

AlexRamey commented on iOS Client/Reducers/MBAppReducer.swift:
```diff
@@ -59,8 +65,27 @@ func devotionReducer(action: Action, state: DevotionState?) -> DevotionState {
+                }
+            }
+            nextState.devotions = Loaded.loaded(data: newDevotions)
+        }
```
since we are dealing with classes (ref types), I think it would work to delete lines 70-79 and just write `nextState.selectedDevotion.read = true`

jdwitten replied:
pretty sure these are all structs though. DevotionState is a struct and devotions are structs

jdwitten replied:
Made this more succinct in the latest commits

AlexRamey commented on iOS Client/Reducers/MBAppReducer.swift:
```diff
@@ -59,8 +65,27 @@ func devotionReducer(action: Action, state: DevotionState?) -> DevotionState {
+            nextState.devotions = Loaded.loaded(data: newDevotions)
+        }
+    case let action as DevotionNotification:
+        nextState.selectedDevotion = action.devotion
```
maybe we should add `nextState.selectedDevotion.read = true` here as well ?

jdwitten replied:
Yes this is a good catch definitely need it here

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -38,6 +38,14 @@ class MBStore: NSObject {
+        } catch {
+            return []
+        }
+    }
```
very simple. I like it.

AlexRamey replied:
the read helper function really shines here

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -78,12 +86,26 @@ class MBStore: NSObject {
+        self.client.getJSONFile(name: "devotions") { data, error in
+            if error == nil, let data = data {
+                do {
+                    //We got some data now parse
```
hells yeah we do

AlexRamey replied:
could use a whitespace between second slash and We. this should be a lint warning if it's not

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -78,12 +86,26 @@ class MBStore: NSObject {
+            } else {
+                //Failed so complete with no devotions
+                completion(nil, error)
+            }
```
very simple. I like it

AlexRamey replied:
the parse helper function really shines here

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -96,51 +118,47 @@ class MBStore: NSObject {
+        try self.save(devotions, "devotions")
+    }
+    
+    private func read<T: Codable>(fromPath: String, _ resource: T.Type) throws -> T {
```
super minor, but if we want to get fancy I think we could name this parameter like `fromPath path:String` (can't remember the exact syntax) so callers get intuitive `fromPath` view of the param name but then internally this function can refer to it as just `path`, which makes more sense when calling urlPackage(forPath: path)

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -96,51 +118,47 @@ class MBStore: NSObject {
+    }
+    
+    private func read<T: Codable>(fromPath: String, _ resource: T.Type) throws -> T {
+        let (manager, path, _) = try urlPackage(forPath: "\(fromPath)")
```
if fromPath is already a string, then why do string interpolation?

jdwitten replied:
Ah just left over from earlier we don't need to do string interpolation any more

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -96,51 +118,47 @@ class MBStore: NSObject {
+        }
+    }
+    
+    private func urlPackage(forPath: String) throws -> (FileManager, String, URL) {
```
same thing here, could expose forPath to callers and name the parameter just path for internal usuage

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -96,51 +118,47 @@ class MBStore: NSObject {
+            let (_, _, pathUrl) = try self.urlPackage(forPath: path)
+            let encoder = JSONEncoder()
+            let encodedData = try encoder.encode(data)
+            try encodedData.write(to: pathUrl, options: [.atomic])
```
nice usage of `.atomic`

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -233,3 +251,10 @@ class MBStore: NSObject {
+    case writeError(msg: String)
+    case parseError(msg: String)
+    case urlError(msg: String)
+}
```
nice

AlexRamey reviewed (changes requested):
Some very nice improvements. I think this is starting to take the shape we want. 

I went through it very thoroughly this time, now that the structure is more in place. 36 comments, but most should be pretty quick to address.

Commit:
Utilize new synchronous methods when scheduling notifications

Commit:
refactor devotion cell formatting

Commit:
refactor reducer and store methods

AlexRamey commented on iOS Client/DevotionsController/DevotionsCoordinator.swift:
```diff
@@ -43,4 +84,14 @@ class DevotionsCoordinator: NSObject, Coordinator, StoreSubscriber {
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+        let devotions = MBStore().getDevotions()
+        let date = Date()
+        if let devotion = devotions.first(where: { Formatters.devotionDateFormatter.date(from: $0.date) == date}) {
```
This is a little dangerous. You're betting that the current date (line 91) exactly matches the devotion's date. Maybe it would be better to convert the current `date` to a string in YYYY-MM-DD format and then do string comparison

AlexRamey replied:
a moot point, but you could use the new lazy `day` property on devotion instead of doing the Formatters call here . . .

AlexRamey replied:
`$0.day == date` would be equivalent; but, again, I think we should do string comparison of YYYY-MM-DD instead of date comparison

jdwitten replied:
Yeah I think you're right we should compare the date strings not the actually date objects

AlexRamey commented on iOS Client/Reducers/MBAppReducer.swift:
```diff
@@ -59,8 +65,27 @@ func devotionReducer(action: Action, state: DevotionState?) -> DevotionState {
+        if case Loaded.loaded(data: var devotions) = nextState.devotions,
+            let index = devotions.index(where: { $0.date == action.devotion.date }) {
+            devotions[index].read = true
+            nextState.devotions = Loaded.loaded(data: devotions)
```
`lines 69-73` and `76-80` are identical. We need to factor this out into a helper of some sort

AlexRamey replied:
Or, equivalently, why do we need the DevotionNotification action? Couldn't we just re-use SelectedDevotion and long as you add `nextState.selectedTab = .devotions` to the navigation reducer for SelectedDevotion action?

jdwitten replied:
I like the second option I can do that

AlexRamey commented on iOS Client/DevotionsController/MBDevotionsViewController.swift:
```diff
@@ -64,5 +72,17 @@ class MBDevotionsViewController: UIViewController, StoreSubscriber, UITableViewD
+            //There was an error saving devotion as read so reverse
+            print("Error marking devotion as read")
+            devotions[indexPath.row].read = false
+            MBStore.sharedStore.dispatch(LoadedDevotions(devotions: .loaded(data: devotions)))
```
Works for now; in a multi-threaded setting we'd want to be careful about clobbering all the devotions in the state with this view controller's local copy, b/c something could change in between the dispatch selected devotion action and this . . . I guess the solution would be an additive only approach where we dispatch a new action called UnselectDevotion or something

AlexRamey commented on iOS Client/Models/MBDevotion.swift:
```diff
@@ -9,9 +9,64 @@
+        self.read = read
+    }
+    
+    var devotion: MBDevotion {
```
where is this used? We should never really go back once we have our domain model.

jdwitten replied:
I don't think I need this I can remove

AlexRamey commented on iOS Client/Utilities/Formatters.swift:
```diff
@@ -0,0 +1,26 @@
+
+import Foundation
+
+struct Formatters {
```
double-check me, but now that we're encapsulating a lot of this work in the LoadDevotion struct, couldn't we remove this file all-together and do these  as private things inside of LoadedDevotion struct? 

(Looks like calendar was already moved over)

1 good use of this file might be a helper that goes from Date to String in YYYY-MM-DD format since that's what we may use for comparison in the notification delegate.

jdwitten replied:
I think in the future if we needed some formatters that were shared between a couple screens or models than we could use this file. I'm ok with having all of this in LoadedDevotion for now

AlexRamey commented on iOS Client/DevotionsController/DevotionDetailViewController.swift:
```diff
@@ -0,0 +1,65 @@
+    func newState(state: MBAppState) {
+        if let devotion = state.devotionState.selectedDevotion,
+            let date = Formatters.devotionDateFormatter.date(from: devotion.date),
+            let monthInt = Formatters.calendar?.component(.month, from: date),
```
there are now helpers for this stuff in LoadedDevotion struct you can use here

Commit:
Remove DevotionNotification action, remove Formatters file, and add replace(devotion: LoadedDevotion) in Store

Commit:
pull month component instead of day for formattedMonth

Commit:
Add an UnreadDevotion function in case there is a failure writing to disk

AlexRamey commented on iOS Client/State/MBState.swift:
```diff
@@ -46,11 +46,19 @@ struct MBArticleState: ArticleState {
 protocol DevotionState {
     var devotions: Loaded<[LoadedDevotion]> { get set }
     var selectedDevotion: LoadedDevotion? { get set }
+    mutating func mark(devotion: LoadedDevotion, asRead read: Bool)
```
nice use of mutating to get ref type semantics

AlexRamey commented on iOS Client/Reducers/MBAppReducer.swift:
```diff
@@ -59,8 +63,18 @@ func devotionReducer(action: Action, state: DevotionState?) -> DevotionState {
+        nextState.selectedDevotion = action.devotion
+        nextState.mark(devotion: action.devotion, asRead: true)
+    case let action as UnreadDevotion:
+        nextState.mark(devotion: action.devotion, asRead: false)
```
really good man

AlexRamey commented on iOS Client/Store/MBStore.swift:
```diff
@@ -96,51 +118,55 @@ class MBStore: NSObject {
+        let markedDevotions = devotions.map { oldDevotion in
+            oldDevotion.date == devotion.date ? devotion : oldDevotion
+        }
+        try self.save(markedDevotions, forPath: "devotions")
```
nice

AlexRamey reviewed (dismissed):
Very nice. This code looks good to me. Look forward to playing with it. I like the idea about doing a middleware to persist the devotions (and maybe to initially read them?) but this is good enough for our purposes. Good work. Let's get this merged and I can hopefully make so contributions to the articles screen this week. DZ wants to get lunch Friday ;)

Commit:
Merge branch 'master' into feature/devotionDetail

AlexRamey reviewed (approved):
merged
